metrics/graphite/graphite.go

```diff
@@ -0,0 +1,295 @@
+	counters   []*counter
+	histograms []*windowedHistogram
+	gauges     []*gauge
+}
+
+// NewEmitter will return an Emitter that will prefix all
+// metrics names with the given prefix. Once started, it will attempt to create
+// a TCP connection with the given address and periodically post metrics to the
+// connection in a Graphite-compatible format.
+func NewEmitter(addr *net.TCPAddr, prefix string) Emitter {
+	e := &emitter{
+		addr, prefix, &sync.Mutex{},
+		[]*counter{}, []*windowedHistogram{}, []*gauge{},
+	}
+
+	return e
+}
+
+// NewCounter returns a Counter whose value will be periodically emitted in
+// a Graphite-compatible format once the Emitter is started. Fields are ignored.
+func (e *emitter) NewCounter(name string) metrics.Counter {
+	// only one flush at a time
+	e.metricMu.Lock()
+	defer e.metricMu.Unlock()
+	c := &counter{name, 0}
+	e.counters = append(e.counters, c)
+	return c
+}
+
+// NewHistogram is taken from http://github.com/codahale/metrics. It returns a
+// windowed HDR histogram which drops data older than five minutes.
+//
+// The histogram exposes metrics for each passed quantile as gauges. Quantiles
+// should be integers in the range 1..99. The gauge names are assigned by
+// using the passed name as a prefix and appending "_pNN" e.g. "_p50".
+//
+// The values of this histogram will be periodically emitted in a Graphite-compatible
+// format once the Emitter is started. Fields are ignored.
+func (e *emitter) NewHistogram(name string, minValue, maxValue int64, sigfigs int, quantiles ...int) metrics.Histogram {
+	// only one flush at a time
+	e.metricMu.Lock()
+	defer e.metricMu.Unlock()
+
+	gauges := map[int]metrics.Gauge{}
+	for _, quantile := range quantiles {
+		if quantile <= 0 || quantile >= 100 {
+			panic(fmt.Sprintf("invalid quantile %d", quantile))
+		}
+		gauges[quantile] = e.gauge(fmt.Sprintf("%s_p%02d", name, quantile))
+	}
+	h := newWindowedHistogram(name, minValue, maxValue, sigfigs, gauges)
+	e.histograms = append(e.histograms, h)
+	return h
+}
+
+// NewTimeHistogram returns a TimeHistogram wrapper around the windowed
+// HDR histrogram provided by this package.
+func (e *emitter) NewTimeHistogram(name string, unit time.Duration, minValue, maxValue int64, sigfigs int, quantiles ...int) metrics.TimeHistogram {
+	h := e.NewHistogram(name, minValue, maxValue, sigfigs, quantiles...)
+	return metrics.NewTimeHistogram(unit, h)
+}
+
+// NewGauge returns a Gauge whose value will be periodically emitted in
+// a Graphite-compatible format once the Emitter is started. Fields are ignored.
+func (e *emitter) NewGauge(name string) metrics.Gauge {
+	// only one flush at a time
+	e.metricMu.Lock()
+	defer e.metricMu.Unlock()
+	return e.gauge(name)
+}
+
+func (e *emitter) gauge(name string) metrics.Gauge {
+	g := &gauge{name, 0}
+	e.gauges = append(e.gauges, g)
+	return g
+}
+
+// Start will kick off a background goroutine to
+// call Flush once every interval.
+func (e *emitter) Start(interval time.Duration) {
```

Comment:
Good catch. `Start` called twice at this state would cause folks pain. I'll wrap the guts in a `sync.Once`.

`Stop` also sounds like a good idea as we could have it call `Flush` one last time and then close the current connection.